irohad/torii/processor/impl/transaction_processor_impl.cpp

```diff
@@ -15,106 +15,123 @@
+                  hash(tx).to_string(), Status::STATELESS_VALIDATION_SUCCESS}));
         }
       });
 
       // move commited txs from proposal to candidate map
-      pcs_->on_commit().subscribe([this](
-          rxcpp::observable<model::Block> blocks) {
+      pcs_->on_commit().subscribe([this](auto blocks) {
         blocks.subscribe(
             // on next..
-            [this](model::Block block) {
-              for (const auto &tx : block.transactions) {
-                if (this->proposal_set_.count(hash(tx).to_string())) {
-                  proposal_set_.erase(hash(tx).to_string());
-                  candidate_set_.insert(hash(tx).to_string());
-                  TransactionResponse response;
-                  response.tx_hash = hash(tx).to_string();
-                  response.current_status =
-                      model::TransactionResponse::STATEFUL_VALIDATION_SUCCESS;
-                  notifier_.get_subscriber().on_next(
-                      std::make_shared<model::TransactionResponse>(response));
-                }
-              }
+            [this](auto block) {
+              const auto in_proposal = [this](const auto &tx) {
+                return this->proposal_set_.count(hash(tx).to_string());
+              };
+              boost::for_each(
+                  block.transactions | boost::adaptors::filtered(in_proposal),
+                  [this](auto &t) { return this->notify_success(t); });
             },
             // on complete
             [this]() {
-              for (auto tx_hash : proposal_set_) {
-                TransactionResponse response;
-                response.tx_hash = tx_hash;
-                response.current_status =
-                    TransactionResponse::STATEFUL_VALIDATION_FAILED;
-                notifier_.get_subscriber().on_next(
-                    std::make_shared<model::TransactionResponse>(response));
-              }
-              proposal_set_.clear();
-
-              for (auto tx_hash : candidate_set_) {
-                TransactionResponse response;
-                response.tx_hash = tx_hash;
-                response.current_status = TransactionResponse::COMMITTED;
-                notifier_.get_subscriber().on_next(
-                    std::make_shared<model::TransactionResponse>(response));
-              }
-              candidate_set_.clear();
+              boost::for_each(this->proposal_set_,
+                              [this](auto &t) { return this->notify_fail(t); });
+              this->proposal_set_.clear();
+              boost::for_each(this->candidate_set_, [this](auto &t) {
+                return this->notify_commit(t);
+              });
+              this->candidate_set_.clear();
             });
       });
+
+      mst_proc_->onPreparedTransactions().subscribe(
+          [this](auto tx) { return this->transactionHandle(tx); });
+      mst_proc_->onExpiredTransactions().subscribe(
```

Comment:
`STATEFUL_VALIDATION_FAILED` does not seem like an appropriate response for this case, since the transaction did not pass through `Simulator` yet. `STATELESS_VALIDATION_FAILED` seems better to me, because the transaction just failed to accumulate enough signatures, whether the signatures correspond to creator or not.

Reply:
Agree, maybe introduce `MST_VALIDATION_FALED`? For me it seems pretty useful to distinguish this errors

Reply:
Agreed on the new `EXPIRED` status